# Tarea - Programación lineal
## Leonardo Rodriguez Hernandez 18-02-2024

### Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----|----|----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo por gramo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, minimizando el costo total. Para simplificar la notación use las siguientes variables: 

+ $x_1:$ Gramos de pollo  
+ $x_2:$ Gramos de carne  
+ $x_3:$ Gramos de cordero  
+ $x_4:$ Gramos de arroz  
+ $x_5:$ Gramos de trigo  
+ $x_6:$ Gramos de gel

La tarea consiste en plantear el problemade programación lineal que permita satisfacer las necesidades alimenticias del gato al tiempo que minimice el costo total y resolverlo con `linprog`.

**Criterio de optimizacion**  
La comida de gato mas barata  

**Restricciones**  
$0.10x_1 + 0.20x_2 + 0.15x_3 + 0.04x_5 \geq 8$  
$0.08x_1 + 0.10x_2 + 0.11x_3 + 0.01x_4 + 0.01x_5 \geq 6$  
$0.001x_1 + 0.005x_2 + 0.005x_3 + 0.10x_4 + 0.15x_5 \leq 2$  
$0.002x_1 + 0.005x_2 + 0.007x_3 + 0.002x_4 + 0.008x_5 \leq 0.4$

**Funcion**  
$0.013x_1 + 0.008x_2 + 0.01x_3 + 0.002x_4 + 0.005x_5 + 0.001x_6$



In [7]:
import numpy as np
c= np.array([0.013,
             0.008,
            0.01,
            0.002,
            0.005,
            0.001])
Aub = np.array([
    [-0.10, -0.20, -0.15, 0 , -0.04, 0],
    [-0.08, -0.10, -0.11, -0.01, -0.01, 0],
    [0.001, 0.005, 0.005, 0.10, 0.15, 0],
    [0.002, 0.005, 0.007, 0.002, 0.008, 0],
    [-1,-1,-1,-1,-1,-1]
])
bub = np.array([-8,
                -6,
                2,
                0.4,
                -100])
import scipy.optimize as opt
opt.linprog(c, Aub, bub)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 0.52
              x: [ 0.000e+00  6.000e+01  0.000e+00  0.000e+00  0.000e+00
                   4.000e+01]
            nit: 2
          lower:  residual: [ 0.000e+00  6.000e+01  0.000e+00  0.000e+00
                              0.000e+00  4.000e+01]
                 marginals: [ 6.400e-03  0.000e+00  1.300e-03  3.000e-04
                              3.300e-03  0.000e+00]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 4.000e+00  0.000e+00  1.700e+00  1.000e-01
                              0.000e+00]
                 marginals: [-0.000e+00 -7.000e-02 -0.000e

**Respuesta**  
Lo mas barato es de 0.52 u.m por cada 100 gramos, cumpliendo con los requisitos.  
La dieta consiste de 60 gramos de carne y 40 gramos de gel.

### Problema de distribución de energía eléctrica

La Comisión Federal de Electricidad **(CFE)** dispone de cuatro plantas de generación para satisfacer la demanda diaria eléctrica en cuatro ciudades, Guadalajara, León y Morelia. Las plantas $1$, $2$ y $3$ pueden satisfacer $80$, $40$ y $60$ millones de KW al día respectivamente. Las necesidades de las ciudades de Guadalajara, León y Morelia son de $70$, $40$ y $70$ millones de Kw al día respectivamente. 


Los costos asociados al envío de suministro energético por cada millón de Kw entre cada planta y cada ciudad son los registrados en la siguiente tabla. 

-|Guadalajara|León|Morelia
:----|----|----|----
Planta 1|5|2|7
Planta 2|3|6|6
Planta 3|6|1|2

Y por último, las restricciones del problema, van a estar dadas por las capacidades de oferta y demanda de cada planta (en millones de KW) y cada ciudad.

Para simplificar la notación use las siguientes variables:

+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades de todas las ciudades al tiempo que minimice los costos asociados a la distribución y resolverlo con `linprog`.

**Criterio de optimizacion**  
Minimizar los costos de distribucion

**Restricciones de oferta**  
$x_1 + x_2 + x_3 \leq 80$  
$x_4 + x_5 + x_6 \leq 40$  
$x_7 + x_8 + x_9 \leq 60$  
**Restricciones de demanda**  
$x_1 + x_4 + x_7 = 70$  
$x_2 + x_5 + x_8 = 40$  
$x_3 + x_6 + x_9 = 70$  

**Funcion de distribucion**  
$5x_1 + 2x_2 + 7x_3 + 3x_4 + 6x_5 + 6x_6 + 6x_7 + x_8 + 2x_9 $

In [8]:
import numpy as np
c= np.array([5,2,7,3,6,6,6,1,2])

Aub = np.array([
    [1,1,1,0,0,0,0,0,0], [0,0,0,1,1,1,0,0,0],[0,0,0,0,0,0,1,1,1]
])
bub = np.array([
                [80,
                40,
                60]
])
Aeq = ([
    [1,0,0,1,0,0,1,0,0],
       [0,1,0,0,1,0,0,1,0],
       [0,0,1,0,0,1,0,0,1]
])
Beq = ([
    [70],
    [40], 
    [70]
])
import scipy.optimize as opt
opt.linprog(c,Aub,bub,Aeq,Beq)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 540.0
              x: [ 3.000e+01  4.000e+01  1.000e+01  4.000e+01  0.000e+00
                   0.000e+00  0.000e+00  0.000e+00  6.000e+01]
            nit: 6
          lower:  residual: [ 3.000e+01  4.000e+01  1.000e+01  4.000e+01
                              0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              6.000e+01]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              6.000e+00  1.000e+00  6.000e+00  4.000e+00
                              0.000e+00]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf        inf        inf
                                    inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00  0.000e+00  0.0

**Analisis de resultado**  
La planta 1 suministra 30 millones de KW a Guadalajara, 40 millones a Leon y 10 millones a morelia.  
La planta 2 suministra 40 millones de KW a Guadalajara solamente.  
La planta 3 suminsitra 60 millones de KW a morelia solamente.  
De esta manera, el costo de envio es de 540 u.m.